### PR TITLE
Move coinbase & coinstake to P2PKH

### PR DESCRIPTION
**Pull request by furszy**
Moved the last remaining P2PK scripts to P2PKH (coinbase and the staking output to the current P2PKH standard).

The main argument used for the migration is that we are already using it as a standard everywhere in the sources (same as upstream) and has an increased, theorical, security due the pub key not being revealed initially (scriptsig reveals it).

### CODE REVIEW

**Comment by random-zebra** [comment]
This is a consensus change.
If not properly handled with a network upgrade, I think it will have nasty consequences, probably forking the chain.

A client updated with this code, staking a P2PKH utxo, will produce a P2PKH coinstake output, and old clients will not be able to verify the block signature (they will try to get the public key from the coinstake output, which in this case will have only the pubkey hash).
New clients, on the other hand, being able to grab the public key from the coinstake input scriptSig, will be able to verify the block signature and accept the block.
This might cause a chain split, or, at least, make it so new clients will be unable to stake "new" utxos (they will be able to stake only the outputs of previous coinstakes, as those are P2PK and remain so).

Further, enforcing `scriptPubKey = scriptPubKeyKernel` for all scripts, is another consensus change.
An old-client staker might theoretically send the stake to a P2PK script different from the one being spent, and use the corresponding key to sign a block. New clients will verify the block signature with a wrong public key (the one taken from the input), thus failing `CheckBlockSignature` and rejecting the block, while the peers running the old code will accept it.

That said, I agree conceptually with the change (although it'll have to wait for 5.0) for the space savings (the privacy/security gain is kinda pointless, as we are spending the same script in the same tx, thus revealing the public key since the creation of the new output anyway).

**Comment by furszy** [comment]
yep 👍 , for 5.0 will be.

**Comment by furszy** [comment]
Updated it with the v5 upgrade check.

**Comment by furszy** [comment]
> It seems that the upgrade logic has been enforced with a block version check.
But the CBlock::CURRENT_VERSION is not bumped here and it sits at 7 (even after the network upgrade).

> The variable onlyP2PK , defined in CreateCoinStake, depends on the chain height (and not on the block version). So, with this code, after the v5 NU, clients would create blocks with P2PKH coinstakes, but still with version 7, and thus those would be rejected by other peers because CheckBlockSignature would fail.

Yeah ok, this PR was another extraction from the Sapling work. The block version bump is due the new Sapling notes merkle tree hash stored on the header. It's not here, neither we are enforcing block v8 at all on this PR. It's done later on the Sapling work.

 Will add the enforcement height check. The idea there was try to prevent the height check as it makes the code ugly. In other scenarios (if we wouldn't be doing a hardfork for Sapling anyway), we could had made this upgrade as a soft fork.

About the test, we spoke about it during the last meeting. Currently, the miner unit tests are not connected and our miner code needs a major refactoring (adapting the `BlockAssembler` class and cleaning up the structure -- btc 7598 path --). Not a straightforward change there to get into this PR.
What can do is add a functional test for now but.. damn that it will just be another slow test.

**Comment by furszy** [comment]
Actually, there are other issues making the height enforcement change.

As the code is now, we cannot add the enforcement height check there. `CheckBlockSignature` is called at the beginning of `ProcessNewBlock` and is processed without tip contextual data (up until that point, we don't know the height of the block, nor if the previous block is already on disk -- if it's not on disk, it will not be processed and the node will request sync from the previous one first).

Will think more about this topic. We shouldn't move the signature check to be a tip contextual check, it needs to be done before even try to look for the previous block on disk to detect fake blocks/chains and not request them.

**Comment by furszy** [comment]
PR updated with an unit test covering the pre/post enforcement block signature verification, added documentation in `checkBlockSignature` function and a functional test validating the chain movement after the upgrade activation.

**Comment by furszy** [comment]
as an extra note, have rebased the PR on top of #1747 to not duplicate the v5 activation height set for regtest. 
We need to merge that one first, then will rebase this one again.

**Comment by furszy** [comment]
done, rebased

File: src/blocksignature.cpp [hunk 1]
```diff
@@ -62,14 +62,34 @@ bool CheckBlockSignature(const CBlock& block)
...
+            } else {
+                int start = 1 + (int) *txin.scriptSig.begin(); // skip sig
+                pubkey = CPubKey(txin.scriptSig.begin()+start+1, txin.scriptSig.end());
+            }
         } else if (whichType == TX_COLDSTAKE) {
             // pick the public key from the P2CS input
             const CTxIn& txin = block.vtx[1].vin[0];
             int start = 1 + (int) *txin.scriptSig.begin(); // skip sig
-            start += 1 + (int) *(txin.scriptSig.begin()+start); // skip flag
+            if (whichType == TX_COLDSTAKE) start += 1 + (int) *(txin.scriptSig.begin()+start); // skip flag
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/blocksignature.cpp]: good catch, it was a leftover from an intermediary commit.

**Comment by furszy** [comment]
Updated per review.

**Pull request merged**